server/project_member.go

```diff
@@ -2,18 +2,218 @@ package server
+		for _, projectMember := range vcsProjectMemberList {
+			if vcs.Type != projectMember.RoleProvider {
+				return echo.NewHTTPError(http.StatusInternalServerError, fmt.Sprintf("Invalid role provider, expected: %v, got: %v", vcs.Type, projectMember.RoleProvider)).SetInternal(err)
+			}
+
+			findPrincipal := &api.PrincipalFind{Email: &projectMember.Email}
+			principal, err := s.PrincipalService.FindPrincipal(ctx, findPrincipal)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to fetch principal info").SetInternal(err)
+			}
+			if principal == nil { // try to create principal
+				signupInfo := &api.Signup{
+					Name:  projectMember.Name,
+					Email: projectMember.Email,
+					// Principal created via this method would have no chance to set their password.
+					// To prevent potential security issues, we use random string to set up her password.
+					// This is another safety measure since we already disallow user login via password
+					// if the principal uses external auth provider
+					Password: common.RandomString(20),
+				}
+				createdPrincipal, httpErr := trySignup(ctx, s, signupInfo, c.Get(getPrincipalIDContextKey()).(int))
+				if httpErr != nil {
+					return httpErr
+				}
+				principal = createdPrincipal
+			}
+
+			providerPayload := &api.ProjectRoleProviderPayload{
+				VCSRole:    projectMember.VCSRole,
+				LastSyncTs: time.Now().UTC().Unix(),
+			}
+			providerPayloadBytes, err := json.Marshal(providerPayload)
+			if err != nil {
+				return echo.NewHTTPError(http.StatusInternalServerError, "Failed to marshal providerPayload").SetInternal(err)
+			}
+			createProjectMember := &api.ProjectMemberCreate{
+				ProjectID:    projectID,
+				CreatorID:    c.Get(getPrincipalIDContextKey()).(int),
+				PrincipalID:  principal.ID,
+				Role:         projectMember.Role,
+				RoleProvider: api.ProjectRoleProvider(projectMember.RoleProvider),
+				Payload:      string(providerPayloadBytes),
+			}
+			createList = append(createList, createProjectMember)
+		}
+
+		createdMemberList, deletedMemberList, err := s.ProjectMemberService.SetProjectMember(ctx, projectID, c.Get(getPrincipalIDContextKey()).(int), createList)
+		if err != nil {
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to sync project member from VCS").SetInternal(err)
+		}
+
+		createdIDMemberMap := make(map[int]*api.ProjectMember)
+		for _, createdMember := range createdMemberList {
+			createdIDMemberMap[createdMember.PrincipalID] = createdMember
+		}
+		deletedIDMemberMap := make(map[int]*api.ProjectMember)
+		for _, deletedMember := range deletedMemberList {
+			deletedIDMemberMap[deletedMember.PrincipalID] = deletedMember
+		}
+
+		// create ROLE CREATE/ MEMBER UPDATE activity
+		for id, createdMember := range createdIDMemberMap {
+			if deletedMember, ok := deletedIDMemberMap[id]; ok {
+				// if the same member exist before, we will create a ROLE UPDATE activity
```

Comment:
This comment doesn't correspond to the following if.

The following if suggests we will skip creating the activity if nothing changed.